src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,110 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int max_key_len = (strlen(format) + strlen(tag) + 1);
+    char tag_idx[9];
+    char open[2] = "[";
+    char close[2] = "]";
+    char *tag_token = NULL;
+    char tag_descriptor[10] = "$TAG";
+    char datestamp[max_key_len];
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    char *tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    tmp_tag = flb_malloc(strlen(tag) + 1);
+    s3_key = flb_sds_create_size(max_key_len);
+    if (!s3_key) {
+        flb_errno();
+        return NULL;
+    }
+    strcpy(s3_key, format);
+    strcpy(tmp_tag, tag);
+
+    /* Finds all occurences of $TAG[*] and replaces then with the right token from tag.*/
+    tag_token = strtok(tmp_tag, tag_delimiter);
+    
+    while(tag_token != NULL) {
+        char tmp[20] = "";
+        sprintf(tag_idx, "%d", i);
+        strcat(tmp, tag_descriptor);
+        strcat(tmp, open);
+        strcat(tmp, tag_idx);
+        strcat(tmp, close);
```

Comment:
small nit/note: we run Fluent Bit code through an automated C code scanning tool before releases, I have a feeling that this might get flagged and it will say that strncat is preferred since it is safer and is considered the best practice.

Also... in fluent bit we'd probably just do:
```
tmp = flb_sds_create_size(10);
// check alloc...
flb_sds_printf(&buf, TAG_PART_FMT, i);
...
```

With this at the top of the file:
```
#define TAG_PART_FMT      "$TAG[%d]"
```

Reply:
(not sure if that sds_printf line is right- check the current code for other uses of it to see)

Reply:
tmp = flb_sds_printf(&tmp...

Reply:
Sorry: tmp = flb_sds_printf(&buf

Reply:
Find usages in the code: `grep -nr 'flb_sds_printf' .`